File: frontend/src/types/Book.ts
export type ReadingStatus = 'UNREAD' | 'READING' | 'COMPLETED';

export interface Book {
  id: number;
  title: string;
  author: string;
  category: string;
  status: ReadingStatus;
  memo?: string;
  registeredAt: string;
  completedAt?: string;
}

export const ReadingStatusLabels: Record<ReadingStatus, string> = {
  UNREAD: '未読',
  READING: '読書中',
  COMPLETED: '読了'
};

export interface BookFormData {
  title: string;
  author: string;
  category: string;
  status: ReadingStatus;
  memo: string;
}

export interface SearchFilters {
  title: string;
  author: string;
  category: string;
  status: ReadingStatus | '';
}

export type SortOption = 
  | 'registered_desc' 
  | 'registered_asc' 
  | 'completed_desc' 
  | 'completed_asc' 
  | 'title' 
  | 'author';

export const SortOptions = [
  { value: 'registered_desc', label: '登録日（新しい順）' },
  { value: 'registered_asc', label: '登録日（古い順）' },
  { value: 'completed_desc', label: '読了日（新しい順）' },
  { value: 'completed_asc', label: '読了日（古い順）' },
  { value: 'title', label: 'タイトル順' },
  { value: 'author', label: '著者順' }
];

File: backend/src/main/java/com/rivel/bookmanager/Entity/Book.java
package com.rivel.bookmanager.Entity;

import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import java.time.LocalDateTime;

@Entity
@Table(name = "books")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Book {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(nullable = false)
    private String title;
    
    @Column(nullable = false)
    private String author;
    
    @Column(nullable = false)
    private String category;
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ReadingStatus status;
    
    @Column(columnDefinition = "TEXT")
    private String memo;
    
    @Column(nullable = false)
    private LocalDateTime registeredAt;
    
    private LocalDateTime completedAt;
    
    public enum ReadingStatus {
        UNREAD("未読"),
        READING("読書中"),
        COMPLETED("読了");
        
        private final String displayName;
        
        ReadingStatus(String displayName) {
            this.displayName = displayName;
        }
        
        public String getDisplayName() {
            return displayName;
        }
    }
    
    @PrePersist
    protected void onCreate() {
        registeredAt = LocalDateTime.now();
    }
    
    @PreUpdate
    protected void onUpdate() {
        if (status == ReadingStatus.COMPLETED && completedAt == null) {
            completedAt = LocalDateTime.now();
        } else if (status != ReadingStatus.COMPLETED) {
            completedAt = null;
        }
    }
}
